hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java

```diff
@@ -81,7 +83,7 @@ public class HoodieMergedLogRecordScanner extends AbstractHoodieLogRecordReader
   // A timer for calculating elapsed time in millis
   public final HoodieTimer timer = HoodieTimer.create();
   // Map of compacted/merged records
-  private final ExternalSpillableMap<String, HoodieRecord> records;
+  private final ExternalSpillableMap<HoodieMergeKey, HoodieRecord> records;
```

Comment:
If we make the key type as a `Serializable`, is it possible to eliminate the `HoodieMergeKey` abstraction? Because the new change introduces the performace regression on non-SI scenarios.

Reply:
We did nto want to overload the `HoodieKey`. This was discussed in https://github.com/apache/hudi/pull/10814#pullrequestreview-2003629004

Reply:
I don't think there is a major perf regression on non-SI path. We are creating only new objects, not doing any further compute.

Reply:
> we do a separate class hierarchy and not overload HoodieKey

Not sure about the specific background here, but make the `ExternalSpillableMap` key as `Serializable` does not overload anything?

Reply:
You're right, it does not overload anything. But, if we make it `Serializable`, then we need to make changes at the callers of `HoodieMergedLogRecordScanner.getRecords()` to decide how to change the `Serializable` to String key. A separate abstraction is cleaner as we don't leak any information at layers above.